sdks/java/core/src/main/java/org/apache/beam/sdk/values/PValueBase.java

```diff
@@ -139,7 +141,15 @@ public boolean isFinishedSpecifyingInternal() {
 
   @Override
   public void finishSpecifying() {
-    finishSpecifyingOutput();
+    checkState(
+        finishedSpecifying,
+        "%s must be finished specifying with finishSpecifying(PInput, PTransform) before any "
```

Comment:
Here, `finishSpecifying` should be idempotent. So if the parameterized version has been called, the no-arg version should be a no-op. So this message could be even more specific: "you ought to call the parameterized version".

If this is not the case, I'd say comment and refer to exceptional cases here.